tidy3d/components/field_projection.py

```diff
@@ -429,96 +425,68 @@ def _far_fields_for_surface(
         phi = np.atleast_1d(phi)
 
-        sin_theta = np.sin(theta)
-        cos_theta = np.cos(theta)
-        sin_phi = np.sin(phi)
-        cos_phi = np.cos(phi)
-
-        J = np.zeros((3, len(theta), len(phi)), dtype=complex)
-        M = np.zeros_like(J)
-
-        phase = [None] * 3
         propagation_factor = -1j * AbstractFieldProjectionData.wavenumber(
             medium=medium, frequency=frequency
         )
 
-        def integrate_for_one_theta(i_th: int):
-            """Perform integration for a given theta angle index"""
-
-            for j_ph in np.arange(len(phi)):
-                phase[0] = np.exp(propagation_factor * pts[0] * sin_theta[i_th] * cos_phi[j_ph])
-                phase[1] = np.exp(propagation_factor * pts[1] * sin_theta[i_th] * sin_phi[j_ph])
-                phase[2] = np.exp(propagation_factor * pts[2] * cos_theta[i_th])
-
-                phase_ij = phase[idx_u][:, None] * phase[idx_v][None, :] * phase[idx_w]
-
-                if self.is_2d_simulation:
-                    J[idx_u, i_th, j_ph] = self.integrate_1d(
-                        currents_f[f"E{cmp_1}"].values, phase_ij, pts[idx_int_1d]
-                    )
-                    J[idx_v, i_th, j_ph] = self.integrate_1d(
-                        currents_f[f"E{cmp_2}"].values, phase_ij, pts[idx_int_1d]
-                    )
-
-                    M[idx_u, i_th, j_ph] = self.integrate_1d(
-                        currents_f[f"H{cmp_1}"].values, phase_ij, pts[idx_int_1d]
-                    )
-
-                    M[idx_v, i_th, j_ph] = self.integrate_1d(
-                        currents_f[f"H{cmp_2}"].values, phase_ij, pts[idx_int_1d]
-                    )
-                else:
-                    J[idx_u, i_th, j_ph] = self.integrate_2d(
-                        currents_f[f"E{cmp_1}"].values, phase_ij, pts[idx_u], pts[idx_v]
-                    )
-
-                    J[idx_v, i_th, j_ph] = self.integrate_2d(
-                        currents_f[f"E{cmp_2}"].values, phase_ij, pts[idx_u], pts[idx_v]
-                    )
+        st = np.sin(theta)
+        ct = np.cos(theta)
+        sp = np.sin(phi)
+        cp = np.cos(phi)
+
+        expr = oe.contract_expression(
+            "xtp,ytp,zt,xyz->xyztp",
+            np.exp(np.einsum("i,j,k->ijk", propagation_factor * pts[0], st, cp)),
+            np.exp(np.einsum("i,j,k->ijk", propagation_factor * pts[1], st, sp)),
+            np.exp(np.einsum("i,j->ij", propagation_factor * pts[2], ct)),
+            currents_f[f"E{cmp_1}"].shape,
+            constants=(0, 1, 2),
+            optimize="optimal",
+        )
 
-                    M[idx_u, i_th, j_ph] = self.integrate_2d(
-                        currents_f[f"H{cmp_1}"].values, phase_ij, pts[idx_u], pts[idx_v]
-                    )
+        jm = []
+        for eh_key in (f"E{cmp_1}", f"E{cmp_2}", f"H{cmp_1}", f"H{cmp_2}"):
+            vals = currents_f[eh_key].attrs.get(AUTOGRAD_KEY, currents_f[eh_key].values)
```

Comment:
How does autograd interact with the current computation here?

Reply:
this is related to how the traced values are stored. essentially, they can either be in `AUTOGRAD_KEY` in the `.attrs` dictionary, or they might be stored in the `DataArray`'s `.values` directly. So this line just extracts those. When no autograd is involved, this is equivalent to just calling `currents[eh_key].values`.